File: R/Antagonize.R
#---------------------------------------------------------------------------------------------------#
Antagonize <- function(M, antprob){
  # Following a certain probability, transform interactions from mutualism to antagonisms. 
  #
  # Args:
  #   M = square binary matrix of mutualistic interactions
  #   antprob = probability of a certain link became a antagonistic link
  #
  # Return:
  #   intr = list of mutualistic and antagonistic networks, respectively
  # 
  V = M * 0 # create V matrix with the same size of mat
  P = matrix(runif(dim(M)[1]*dim(M)[2], min = 0, max = 1),  # create matrix of probabilities
              ncol = dim(M)[1], nrow = dim(M)[2])

  M[antprob >= P] = 0 # if the probability is lower or equal than antprob,the link in M in zero
  V[antprob >= P] = 1 # and the link in V is 1
  diag(V) = 0 # diagonal's matrices must be zero

  mats = list(M, V) # create list with the antagonistic and mutualistic matrices
  return(mats) # return a list with the antagonistic and mutualistic matrices
}
#---------------------------------------------------------------------------------------------------#

File: Python/Antagonize.py
def Antagonize(M, antprob):
    ''' Function to antagonize interactions of a adjacency matriz of 
    interactions following a certain probability called antprob.
    Args:
    M: Adjacency matrix of interactions
    antprob: Probabilibty of a link become antagonist
    Return:
    M: Adjacency matrix of interactions with 0's
    V: Adjacency matrix of antagonistic interactions with 1's'''
    
    # import modules
    import numpy as np
    # Generate an empty matrix V with equal dimension of M
    V = M * 0
    # Generate numbers from 0 to 1
    P = np.array(np.random.rand(len(M), len(M)))
    # Off the link if the number generated in P are bigger than antprob
    M[antprob >= P] = 0
    # On the link if the number generated in P are bigger than antprob
    V[antprob >= P] = 1 
    # matrix diagonal must be zero
    np.fill_diagonal(V, 0)
    # Return the matrices
    return M, V